providers/implementations/kdfs/argon2.c

```diff
@@ -0,0 +1,1584 @@
+}
+
+static void kdf_argon2_reset(void *vctx)
+{
+    OSSL_LIB_CTX *libctx;
+    KDF_ARGON2 *ctx;
+    ARGON2_TYPE type;
+
+    ctx = (KDF_ARGON2 *) vctx;
+    type = ctx->type;
+    libctx = ctx->libctx;
+
+    if (ctx->out != NULL)
+        OPENSSL_clear_free(ctx->out, ctx->outlen);
+
+    if (ctx->pwd != NULL)
+        OPENSSL_clear_free(ctx->pwd, ctx->pwdlen);
+
+    if (ctx->salt != NULL)
+        OPENSSL_clear_free(ctx->salt, ctx->saltlen);
+
+    if (ctx->secret != NULL)
+        OPENSSL_clear_free(ctx->secret, ctx->secretlen);
+
+    if (ctx->ad != NULL)
+        OPENSSL_clear_free(ctx->ad, ctx->adlen);
+
+    memset(ctx, 0, sizeof(*ctx));
+    ctx->libctx = libctx;
+    kdf_argon2_init(ctx, type);
+}
+
+static int kdf_argon2_ctx_set_threads(KDF_ARGON2 *ctx, uint32_t threads)
+{
+    if (threads < ARGON2_MIN_THREADS) {
+        ERR_raise_data(ERR_LIB_PROV, PROV_R_INVALID_THREAD_POOL_SIZE,
+                "min threads: %u", ARGON2_MIN_THREADS);
+        return 0;
+    }
+
+    if (threads > ARGON2_MAX_THREADS) {
+        ERR_raise_data(ERR_LIB_PROV, PROV_R_INVALID_THREAD_POOL_SIZE,
+                "max threads: %u", ARGON2_MAX_THREADS);
+        return 0;
+    }
+
+    ctx->threads = threads;
+    return 1;
+}
+
+static int kdf_argon2_ctx_set_lanes(KDF_ARGON2 *ctx, uint32_t lanes)
+{
+    if (lanes > ARGON2_MAX_LANES) {
+        ERR_raise_data(ERR_LIB_PROV, PROV_R_FAILED_TO_SET_PARAMETER,
+                "max lanes: %u", ARGON2_MAX_LANES);
+        return 0;
+    }
+
+    if (lanes < ARGON2_MIN_LANES) {
+        ERR_raise_data(ERR_LIB_PROV, PROV_R_FAILED_TO_SET_PARAMETER,
+                "min lanes: %u", ARGON2_MIN_LANES);
+        return 0;
+    }
+
+    ctx->lanes = lanes;
+    return 1;
+}
+
+static int kdf_argon2_ctx_set_t_cost(KDF_ARGON2 *ctx, uint32_t t_cost)
+{
+    /* ARGON2_MAX_MEMORY == max m_cost value, skip check, enforce type */
+    ossl_static_assert_type_eq(uint32_t, t_cost);
+
+    if (t_cost < ARGON2_MIN_TIME) {
+        ERR_raise_data(ERR_LIB_PROV, PROV_R_INVALID_ITERATION_COUNT,
+                "min: %u", ARGON2_MIN_TIME);
+        return 0;
+    }
+
+    ctx->t_cost = t_cost;
+    return 1;
+}
+
+static int kdf_argon2_ctx_set_m_cost(KDF_ARGON2 *ctx, uint32_t m_cost)
+{
+    /* ARGON2_MAX_MEMORY == max m_cost value, skip check, enforce type */
+    ossl_static_assert_type_eq(uint32_t, m_cost);
+
+    if (m_cost < ARGON2_MIN_MEMORY) {
+        ERR_raise_data(ERR_LIB_PROV, PROV_R_INVALID_MEMORY_SIZE, "min: %u",
+            ARGON2_MIN_MEMORY);
+        return 0;
+    }
+
+    ctx->m_cost = m_cost;
+    return 1;
+}
+
+static int kdf_argon2_ctx_set_out_length(KDF_ARGON2 *ctx, uint32_t outlen)
+{
+    /* ARGON2_MAX_OUT_LENGTH == max outlen value, so upper bounds checks
+     * are always satisfied; to suppress compiler if statement tautology
+     * warnings, these checks are skipped; however, to ensure that these
+     * limits are met and implementation conforming to Argon2 RFC, we need
+     * to fix the type */
```

Comment:
Nit: Please fix the formatting of the multiline comment (beginning and end marker on separate line)